toffy/image_stitching.py

```diff
@@ -0,0 +1,71 @@
+import os
+import shutil
+import math
+import natsort as ns
+import skimage.io as io
+
+from toffy import json_utils
+from ark.utils import data_utils, load_utils, io_utils, misc_utils
+from mibi_bin_tools.io_utils import remove_file_extensions
+
+
+def get_max_img_size(run_dir):
+    """Retrieves the maximum FOV image size listed in the fun file
+        Args:
+            run_dir (str): path to the run directory containing the run json files """
+
+    run_name = os.path.basename(run_dir)
+    run_file_path = os.path.join(run_dir, run_name + '.json')
+
+    # retrieve all pixel width dimensions of the fovs
+    run_data = json_utils.read_json_file(run_file_path)
+    img_sizes = []
+    for fov in run_data['fovs']:
+        img_sizes.append(fov.get('frameSizePixels')['width'])
+
+    # return largest
+    max_img_size = max(img_sizes)
+
+    return max_img_size
+
+
+def stitch_images(tiff_out_dir, run_dir, channels=None):
+    """Creates a new directory containing stitched channel images for the run
+        Args:
+            tiff_out_dir (str): path to the extracted images for the specific run
+            run_dir (str): path to the run directory containing the run json files
+            channels (list): list of channels to produce stitched images for, None will do all """
+
+    # remove old images
+    stitched_dir = os.path.join(tiff_out_dir, 'stitched_images')
+    if os.path.exists(stitched_dir):
+        shutil.rmtree(stitched_dir)
+
+    folders = io_utils.list_folders(tiff_out_dir)
+    if 'intensities' in folders:
+        folders.remove('intensities')
+    folders = ns.natsorted(folders)
+
+    # retrieve all extracted channel names, or verify the list provided
+    if channels is None:
+        channels = remove_file_extensions(io_utils.list_files(
+            dir_name=os.path.join(tiff_out_dir, folders[0]), substrs='.tiff'))
+    else:
+        misc_utils.verify_in_list(channel_inputs=channels, valid_channels=remove_file_extensions(
+            io_utils.list_files(dir_name=os.path.join(tiff_out_dir, folders[0]), substrs='.tiff')))
+
+    # load in and stitch the image data
+    num_cols = math.isqrt(len(folders))
+    max_img_size = get_max_img_size(run_dir)
+    image_data = load_utils.load_imgs_from_tree(tiff_out_dir, fovs=folders, channels=channels,
```

Comment:
The loading and stitching should go in the for loop below as well, that way we're only loading one channels worth of image data at a time. Otherwise this takes forever to run since all of the data has to be loaded.